src/main/java/gregtech/common/EventHandlers.java

```diff
@@ -206,9 +213,92 @@ public static void onLivingEquipmentChangeEvent(LivingEquipmentChangeEvent event
         }
     }
 
+    @SuppressWarnings({ "ConstantValue", "deprecation" })
     @SubscribeEvent
-    @SideOnly(Side.CLIENT)
     public static void onPlayerTick(TickEvent.PlayerTickEvent event) {
+        EntityPlayer player = event.player;
+        if (event.phase == TickEvent.Phase.START && !player.world.isRemote) {
+            IAttributeInstance movementSpeed = player.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED);
+            if (movementSpeed == null) return;
+            AttributeModifier modifier = movementSpeed.getModifier(BlockUtility.WALKING_SPEED_UUID);
+
+            double speedBonus;
+            if (!player.onGround || player.isInWater() || player.isSneaking()) {
+                speedBonus = 0;
+            } else {
+                IBlockState state = player.world.getBlockState(new BlockPos(
+                        player.posX, player.getEntityBoundingBox().minY - 1, player.posZ));
+                speedBonus = BlockUtility.WALKING_SPEED_BONUS.getDouble(state);
+                if (speedBonus == 0 &&
+                        state.getBlock() instanceof IWalkingSpeedBonus walkingSpeedBonus &&
+                        walkingSpeedBonus.getWalkingSpeedBonus() != 1 &&
+                        walkingSpeedBonus.bonusSpeedCondition(player) &&
+                        walkingSpeedBonus.checkApplicableBlocks(state)) {
+                    speedBonus = walkingSpeedBonus.getWalkingSpeedBonus() - 1;
+                }
+            }
+            if (modifier != null) {
+                if (speedBonus == modifier.getAmount()) return;
+                else movementSpeed.removeModifier(BlockUtility.WALKING_SPEED_UUID);
+            } else {
+                if (speedBonus == 0) return;
+            }
```

Comment:
Is there a need to check if the speedBonus is 0 here? We failed to retrieve the movement speed modifier, so nothing can be applied, even if the bonus was not zero

Reply:
No modifier instance means no speed bonus is applied to player at the moment, new modifiers can be applied to the attribute instance. Since 0 speed bonus is pointless to apply, the modifier is just removed entirely when speed bonus of 0 gets applied.